src/wasm/wasm-type.cpp

```diff
@@ -260,24 +324,11 @@ TypeInfo::~TypeInfo() {
 }
 
 bool TypeInfo::operator==(const TypeInfo& other) const {
-  if (kind != other.kind) {
-    return false;
-  }
-  switch (kind) {
-    case TupleKind:
-      return tuple == other.tuple;
-    case RefKind:
-      return ref.heapType == other.ref.heapType &&
-             ref.nullable == other.ref.nullable;
-    case RttKind:
-      return rtt == other.rtt;
-  }
-  WASM_UNREACHABLE("unexpected kind");
+  return TypeStructureEquator().eq(*this, other);
```

Comment:
Is the idea that it is ok to only compare the finite shape because this is after global canonicalization? (if so then a comment maybe could help)

Reply:
Not after global canonicalization, but after shape canonicalization. Will add a comment.